### PR TITLE
torch2flow

### PR DESCRIPTION
**Pull request by zhongshsh**
将 Pytorch 存储的预训练模型转为 OneFlow 模型。
- issue： https://github.com/Oneflow-Inc/oneflow-documentation/issues/457 
- 文档预览： https://zhongshsh.github.io/oneflow-documentation/master/cookies/torch2flow.html

### CODE REVIEW

**Comment by doombeaker** [comment]
在 PR 描述里，最好更详细地描述清楚这个 PR 做了什么。
比如相关联的 ISSUE、预览 URL 地址这类。

File: cn/docs/cookies/torch2flow.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+
+自然语言处理、图像等深度学习领域常使用预训练模型+微调机制来快速解决具体任务，优化器参数等实验信息往往也有存储需求。因此，在使用 OneFlow 进行模型训练时，不可避免地会遇到如何将 Pytorch 存储的预训练模型转为 OneFlow 模型的问题。
+
+
+
+# 转换示例
```
Comment (doombeaker) [inline cn/docs/cookies/torch2flow.md:6]:
```suggestion
## 转换示例

```

`##` 是二级标题。一级标题一般指的是整个文章的标题。下面的也要改

Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md]: 收到

File: cn/docs/cookies/torch2flow.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+```python
+parameters = torch.load(save_file).state_dict()
+
+for key, value in parameters.items():
+    val = value.detach().cpu().numpy()
+    parameters[key] = val
+
+model_flow.load_state_dict(parameters)
+```
```
Comment (doombeaker) [inline cn/docs/cookies/torch2flow.md:49]: 感觉这两段代码可以合到一起。说明也合到一起。

然后操作完成后，重点说明下是用 numpy 作媒介

File: cn/docs/cookies/torch2flow.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+最后，我们加载 Pytorch 存储的模型文件 `model.pth`，并将其初始化到 OneFlow 版本的模型中。
+
+```python
+parameters = torch.load(save_file).state_dict()
+
+for key, value in parameters.items():
+    val = value.detach().cpu().numpy()
+    parameters[key] = val
```
Comment (doombeaker) [inline cn/docs/cookies/torch2flow.md:46]: 在这里加一句打印 key, val。这样就省略掉之后那个验证代码以及对应解释了。

File: cn/docs/cookies/torch2flow.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+model_flow.load_state_dict(parameters)
+```
+
+我们可以通过如下代码检验模型是否加载成功。
```
Comment (doombeaker) [inline cn/docs/cookies/torch2flow.md:51]: 这里到本节结尾感觉都可以试着省去。

File: cn/docs/cookies/torch2flow.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        [0.6272, 0.3728],
+        [0.6275, 0.3725],
+        [0.5735, 0.4265]], dtype=oneflow.float32, grad_fn=<softmax_backward>)
+```
+
+
+
+## 拓展
+
+torchvision 中有许多预训练的图像模型，那么如何将 torchvision 中的预训练模型转为 OneFlow 模型呢？以经典的  AlexNet 为例，转换代码如下所示：
```
Comment (doombeaker) [inline cn/docs/cookies/torch2flow.md:116]: 这里的表述可以变一下，重点要突出的有：

1. 以上是简单例子，但其实只要确保模型对齐，多么复杂的模型都可以转换
2. flowvision 是和 torchvision 对齐的模型库，以此举个实际例子

并且举例完后，是不是再强调下，flowvision 其实也都配了预训练模型，设置 `pretrained=True` 即可，并不是真正需要每个都转换。

File: cn/mkdocs.yml [hunk 6]
```diff
@@ -132,6 +132,7 @@ nav:
     - 实践指南:
       - OneFlow 与 ONNX 交互: cookies/oneflow2onnnx.md
       - 模型部署: cookies/serving.md
+      - 将 Pytorch 存储的预训练模型转为 OneFlow 模型: cookies/torch2flow.md
```
Comment (doombeaker) [inline cn/mkdocs.yml:135]:
```suggestion
      - 将 Pytorch 预训练模型转为 OneFlow 格式: cookies/torch2flow.md
```

File: cn/docs/cookies/torch2flow.md [hunk 7]
```diff
@@ -0,0 +1,144 @@
+
```
[the remark this answers is not in the document]
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:1]: 收到

File: cn/docs/cookies/torch2flow.md [hunk 8]
```diff
@@ -0,0 +1,144 @@
+
+自然语言处理、图像等深度学习领域常使用预训练模型+微调机制来快速解决具体任务，优化器参数等实验信息往往也有存储需求。因此，在使用 OneFlow 进行模型训练时，不可避免地会遇到如何将 Pytorch 存储的预训练模型转为 OneFlow 模型的问题。
```
[the remark this answers is not in the document]
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:2]: 收到

File: cn/docs/cookies/torch2flow.md [hunk 9]
```diff
@@ -0,0 +1,145 @@
+
+自然语言处理、图像等深度学习领域常使用预训练模型+微调机制来快速解决具体任务，优化器参数等实验信息往往也有存储需求。因此，在使用 OneFlow 进行模型训练时，不可避免地会遇到如何将 Pytorch 存储的预训练模型转为 OneFlow 模型的问题。
+
+
+
+## 转换示例
+
+
+将 Pytorch 存储的数据（无论是模型还是变量等等）转换成 OneFlow 的思路是 **使用 Numpy 作为二者的媒介**，具体如下示例所示。
```
[the remark this answers is not in the document]
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:9]: 收到

File: cn/docs/cookies/torch2flow.md [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+model_torch = nn.Sequential(
+    nn.Linear(128, 2), 
+    nn.Softmax()
+)
+
+torch.save(model, save_file)
+```
+
+接着，我们需要定义一个具有 **相同结构** 的 OneFlow 版本的模型。
```
[the remark this answers is not in the document]
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:27]: 收到

File: cn/docs/cookies/torch2flow.md (hunk 3 above)
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:46]: 收到

File: cn/docs/cookies/torch2flow.md (hunk 2 above)
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:49]: 收到

File: cn/docs/cookies/torch2flow.md (hunk 4 above)
Comment (zhongshsh) [reply cn/docs/cookies/torch2flow.md:51]: 收到

File: cn/mkdocs.yml (hunk 6 above)
Comment (zhongshsh) [reply cn/mkdocs.yml:135]: 收到

**Comment by doombeaker** [comment]
修改后不错，感觉已经可以发布了。不过先不着急合并，我协同下另外的同学，在这个PR添加英文版后，再一起合并。

**Review by doombeaker — APPROVED** [review]
